src/services/db/GitFileSystemService.ts

```diff
@@ -0,0 +1,278 @@
+    }
+
+    try {
+      await this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).pull()
+    } catch (error: unknown) {
+      if (error instanceof GitError) {
+        return errAsync(new GitFileSystemError(error.message))
+      }
+
+      return errAsync(error)
+    }
+
+    return okAsync(null)
+  }
+
+  // Creates either directory or file
+  // ResourceDirectoryService used this to create a directory + file at the same time
+  async create() {}
+
+  // Read the contents of a file
+  async read(
+    repoName: string,
+    filePath: string
+  ): Promise<Result<GitFile, GitFileSystemError | unknown>> {
+    // Ensure that the repository is up-to-date first
+    const isUpdated = await this.pull(repoName)
+
+    if (isUpdated.isErr()) {
+      return errAsync(isUpdated.error)
+    }
+
+    // Read the file and get the hash concurrently
+    try {
+      const [contents, sha] = await Promise.all([
+        fs.promises.readFile(
+          `${EFS_VOL_PATH}/${repoName}/${filePath}`,
+          "utf-8"
+        ),
+        this.getGitBlobHash(repoName, filePath),
+      ])
+
+      if (sha.isErr()) {
+        return errAsync(sha.error)
+      }
+
+      const result: GitFile = {
+        contents,
+        sha: sha.value,
+      }
+
+      return okAsync(result)
+    } catch (error: unknown) {
+      if (error instanceof Error) {
+        return errAsync(new GitFileSystemError(error.message))
+      }
+
+      return errAsync(error)
+    }
+  }
+
+  // Read the contents of a directory
+  async listDirectoryContents(
+    repoName: string,
+    directoryPath: string
+  ): Promise<Result<GitDirectoryItem[], GitFileSystemError | unknown>> {
+    // Check that the directory path provided exists and is a directory
+    try {
+      const stats = await fs.promises.stat(
+        `${EFS_VOL_PATH}/${repoName}/${directoryPath}`
+      )
+
+      if (!stats.isDirectory()) {
+        return errAsync(
+          new GitFileSystemError(
+            `Path "${directoryPath}" is not a directory in repo "${repoName}"`
+          )
+        )
+      }
+    } catch (error: unknown) {
+      if (error instanceof Error) {
+        // Likely the path does not exist
+        return errAsync(new GitFileSystemError(error.message))
+      }
+
+      return errAsync(error)
+    }
+
+    // Read the directory contents
+    try {
+      const directoryContents = await fs.promises.readdir(
+        `${EFS_VOL_PATH}/${repoName}/${directoryPath}`,
+        { withFileTypes: true }
+      )
+
+      const result: GitDirectoryItem[] = await Promise.all(
+        directoryContents.map(async (directoryItem) => {
+          const isDirectory = directoryItem.isDirectory()
+          const { name } = directoryItem
+          const path = directoryPath === "" ? name : `${directoryPath}/${name}`
+          const type = isDirectory ? "dir" : "file"
+          const fileHash = await this.getGitBlobHash(repoName, path)
+          const sha = fileHash.isOk() ? fileHash.value : ""
```

Comment:
this appears to only do a single level deep read? we also appear to call `getGitBlobHash` on the directory - is this intended?

Reply:
I think our existing directory implementation (in GithubService) also uses this same single level deep read, so this should be consistent

Reply:
sgtm